rfcs/replace_ssh_with_ssm_for_linux_template/replace_ssh_with_ssm_for_linux_template.md

```diff
@@ -0,0 +1,238 @@
+```
+  EC2SSMRole:
+    Type: 'AWS::IAM::Role'
+    Properties:
+      RoleName: !Join ['-', [Ref: Namespace, 'ec2-ssm-role']]
+      Path: '/'
+      AssumeRolePolicyDocument:
+        Version: '2012-10-17'
+        Statement:
+          - Effect: 'Allow'
+            Principal:
+              AWS:
+                - !Ref ApiHandlerArn
+            Action:
+              - 'sts:AssumeRole'
+      Policies:
+        - PolicyName: !Join ['-', [Ref: Namespace, 'ssm-session-policy']]
+          PolicyDocument:
+            Version: '2012-10-17'
+            Statement:
+              - Effect: 'Allow'
+                Action: 'ssm:StartSession'
+                Resource: !Join 
+                  - ''
+                  - - 'arn:'
+                    - !Ref AWS::Partition
+                    - ':ec2:'
+                    - !Ref AWS::Region
+                    - ':'
+                    - !Ref AWS::AccountId
+                    - ':instance/'
+                    - !Ref EC2Instance
+
+```
+
+We also need to output this role arn through CloudFormation Output as below:
+```
+  WorkspaceSSMRoleArn:
+    Description: IAM role assumed by API Handler to start session
+    Value: !GetAtt EC2SSMRole.Arn
+```
+
+### Workspace Connection Sequence
+We need to refactor the sequence of workspace connection process as below:
+![ssm-design.drawio](./ssm-design.drawio.png)
+
+### Get Connections
+After user clicking "Connection" green button, client side will send ```/api/workspaces/service-catalog/${envId}/connections``` request to server side. Server side will return below content to client side:
+```
+[
+  {
+    "scheme": "https",
+    "id": "id-0",
+    "type": "linux",
+    "name": "Linux Instance",
+    "operation": "list"
+  }
+]
+```
+The response is depended on the cloudformation template output that mentioned in previous section, the logic of ``Get Connections`` is not needed to change.
+
+### Get Detail Information to Connect
+After user clicking "Connect" blue button, client side will send ``/api/workspaces/service-catalog/${envId}/connections/${connectionId}/url`` request to server side to get session manager url with siginToken. During this process, we need to implement as below:
+* Assume role to get temporary AKSK
+* Get temporary signin token
+* Generate session manager url with signin token
+
+1. Assume role:
+The ``getCredentialsForRole()`` method is existing in current source code in addons/addon-base/packages/services/lib/aws/aws-service.js file. We can reuse it to get ``EC2SSMRole``'s AKSK, it will be used for getting signIn token.
+
+
+2. Get temporary signin token
+We can get signIn token through the AKSK, the code snap is like below:
+```
+  const session = {'sessionId': accessKeyId, 'sessionKey': secretAccessKey 'sessionToken': sessionToken};
+
+  const getSigninTokenParams={
+    'Action': 'getSigninToken',
+    'SessionDuration': 43200, // 12 hours
```

Comment:
might be the duration can be exposed by workspace template, you can set a default one(1 hour is recommended) when it's absent

Reply:
Sure, refactored the design doc